### PR TITLE
[Intel GPU] Support RegisterXPU.cpp codegen and compile for the in-tree XPU structured GEMM OPs.

### PR DESCRIPTION
**Pull request by etaf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135320
* #135318
* #139026
* #136742
* __->__ #139025

[Intel GPU] Support RegisterXPU.cpp codegen and compile for the in-tree XPU structured GEMM ops.

Motivation: There are two parts of aten ops for XPU, one is in-tree ops like GEMM related OPs and the other is out-off-tree ops in torch-xpu-ops. For the in-tree part，since Pytorch uses native_functions.yaml registration and is equipped with convenient codegen capabilities, we want to take advantage of these benefits as well.
At the same time, since AOT Inductor also uses native_functions.yaml to generate c shim wrappers, we also need to enable this mechanism for XPU.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
@etaf , pls. refine the PR description to elaborate on why we need to do such changes.

**Review by jansel — APPROVED** [review]

**Comment by etaf** [comment]
@desertfire  Could you please review this PR?

File: torchgen/gen.py [hunk 1]
```diff
@@ -2195,7 +2217,17 @@ def gen_source_files(
...
+            next(
+                (
+                    device
+                    for check, device in dispatch_device_map.items()
+                    if check(dispatch_key)
+                ),
+                "cpu",
+            ),
+            cpu_fm,
+        )
```
Comment (desertfire) [inline torchgen/gen.py:2230]: Repeated enough times that we should write a utility function for it.

File: torchgen/gen.py [hunk 2]
```diff
@@ -2833,6 +2871,19 @@ def main() -> None:
         if DispatchKey.MPS in dispatch_keys:
             del dispatch_keys[dispatch_keys.index(DispatchKey.MPS)]
 
+    xpu_in_whitelist = (
+        options.backend_whitelist and str(DispatchKey.XPU) in options.backend_whitelist
+    )
+    # Only generate RegisterXPU.cpp when there is "--xpu" with torhgen/gen.py
+    # Before this change, torchgen always generates RegisterXPU.cpp for out-of-tree
+    # torch-xpu-ops native_functions.yaml which use --backend_whitelist=XPU and without "--xpu".
+    # After this change is landed, we will add --xpu in torch-xpu-ops and rmmove the check of "xpu_in_whitelist".
```
Comment (desertfire) [inline torchgen/gen.py:2880]: typo: rmmove

**Review by desertfire — APPROVED** [review]

File: torchgen/gen.py (hunk 2 above)
Comment (etaf) [reply torchgen/gen.py]: resolved, thanks.

**Pull request closed without merging**